File: src/models/mongo/flow.py
from typing import List, Optional, Any
from datetime import datetime
from beanie import Document, Indexed, PydanticObjectId
from pydantic import Field
from src.schemas.flow import FlowNode, FlowEdge

class FlowDocument(Document):
    """
    Representação persistente de um Fluxo no MongoDB via Beanie.
    Replaces the 'Flow' C# entity from .NET.
    """
    id: Optional[PydanticObjectId] = Field(None, alias="_id")
    tenant_id: Indexed(str)
    name: str = Field(..., max_length=100)
    description: Optional[str] = None
    
    # Estrutura do Grafo
    nodes: List[FlowNode]
    edges: List[FlowEdge]
    
    # Configurações de Gatilho
    trigger_keywords: List[str] = Field(default_factory=list)
    is_active: bool = True
    
    # Auditoria
    updated_at: datetime = Field(default_factory=datetime.utcnow)
    version: int = 1

    class Settings:
        name = "flows"
        indexes = [
            "tenant_id",
            [("tenant_id", 1), ("name", 1)],
            [("tenant_id", 1), ("is_active", 1)]
        ]

class SessionStateDocument(Document):
    """
    Estado da sessão atual de um usuário em um fluxo específico.
    Controla em qual nó o usuário parou e as variáveis coletadas.
    """
    tenant_id: Indexed(str)
    contact_phone: Indexed(str)
    flow_id: Indexed(str)
    current_node_id: str
    
    # Variáveis coletadas durante o fluxo (ex: nome, idade, intenção)
    variables: dict = Field(default_factory=dict)
    
    # Controle de Tempo e Handover (Sprint 21)
    last_interaction: datetime = Field(default_factory=datetime.utcnow)
    is_completed: bool = False
    is_human_support: bool = False

    class Settings:
        name = "flow_sessions"

File: node-version/src/models/nosql/Flow.js
const mongoose = require('mongoose');

// Schemas embutidos
const flowNodeSchema = new mongoose.Schema({
  id: String,
  type: String,
  data: mongoose.Schema.Types.Mixed,
  position: { x: Number, y: Number }
}, { _id: false });

const flowEdgeSchema = new mongoose.Schema({
  id: String,
  source: String,
  target: String,
  type: String,
  label: String
}, { _id: false });

const flowSchema = new mongoose.Schema({
  tenant_id: { type: String, required: true, index: true },
  name: { type: String, required: true, maxlength: 100 },
  description: { type: String, default: null },
  
  nodes: [flowNodeSchema],
  edges: [flowEdgeSchema],
  
  trigger_keywords: { type: [String], default: [] },
  is_active: { type: Boolean, default: true },
  
  updated_at: { type: Date, default: Date.now },
  version: { type: Number, default: 1 }
}, { collection: 'flows' });

flowSchema.index({ tenant_id: 1, name: 1 });
flowSchema.index({ tenant_id: 1, is_active: 1 });

const Flow = mongoose.model('Flow', flowSchema);

const sessionStateSchema = new mongoose.Schema({
  tenant_id: { type: String, required: true, index: true },
  contact_phone: { type: String, required: true, index: true },
  flow_id: { type: String, required: true, index: true },
  current_node_id: { type: String, required: true },
  
  variables: { type: mongoose.Schema.Types.Mixed, default: {} },
  
  last_interaction: { type: Date, default: Date.now },
  is_completed: { type: Boolean, default: false },
  is_human_support: { type: Boolean, default: false }
}, { collection: 'flow_sessions' });

const SessionState = mongoose.model('SessionState', sessionStateSchema);

module.exports = { Flow, SessionState };
